File: python/12-02/12-02.py
import json

with open('data.txt', 'r') as data_file:
    data = [d for d in data_file.read().split('\n')]

with open('data.json', 'w') as file:
    file.write(json.dumps(data, indent=2))


def part1(data):
    position = {
        'forward': 0,
        'depth': 0
    }
    for move in data:
        direction, distance = move.split(' ')
        distance = int(distance)
        match direction:
            case 'forward':
                position['forward'] += distance
            case 'up':
                position['depth'] -= distance
            case 'down':
                position['depth'] += distance
    
    return position['forward'] * position['depth']

# print(part1(data)) # 1714680


def part2(data):
    position = {
        'aim': 0,
        'forward': 0,
        'depth': 0
    }

    for move in data:
        direction, value = move.split(' ')
        value = int(value)
        match direction:
            case 'forward':
                position['forward'] += value
                position['depth'] += position['aim'] * value
            case 'up':
                position['aim'] -= value
            case 'down':
                position['aim'] += value

    return position['forward'] * position['depth']

# print(part2(data))

File: javascript/12-02/12-02.js
let data = require('./data.json')

const part1 = data => {
  let position = {
    forward: 0,
    depth: 0
  }

  data.forEach(move => {
    let [direction, distance] = move.split(' ')
    distance = parseInt(distance)

    switch (direction) {
      case 'forward':
        position.forward += distance
        break
      case 'up':
        position.depth -= distance
        break
      case 'down':
        position.depth += distance
        break
    }
  })

  return position.depth * position.forward
}

// console.log(part1(data)) // 1714680

const part2 = data => {
  let position = {
    aim: 0,
    forward: 0,
    depth: 0
  }

  data.forEach(move => {
    let [direction, value] = move.split(' ')
    value = parseInt(value)

    switch (direction) {
      case 'forward':
        position.forward += value
        position.depth += position.aim * value
        break
      case 'up':
        position.aim -= value
        break
      case 'down':
        position.aim += value
        break
      default:
        break
    }
  })

  return position.depth * position.forward
}

// console.log(part2(data)) // 1963088820
